### PR TITLE
Improve msan behavior of dbfopen API calls.

### PR DESCRIPTION
**Pull request by slacy**
- If within a method that can fail, return error result when FRead() fails.
    - Otherwise, Convert FRead(..., Size, 1) to FRead(..., 1, Size).
    - Only change places where the return value of FRead() is currently ignored.
    - Fixes another msan issue similar to OSGeo#2970

### CODE REVIEW

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 1]
```diff
@@ -304,16 +304,16 @@ void SHPAPI_CALL
...
 {
-    unsigned char		abyFileHeader[XBASE_FILEHDR_SZ];
 
     if( psDBF->bNoHeader )
         DBFWriteHeader( psDBF );
 
     if( !DBFFlushRecord( psDBF ) )
         return;
 
+    unsigned char abyFileHeader[XBASE_FILEHDR_SZ] = {0};
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:314]: when merged back in shapelib, this is still intended to be C89 I think, so the declaration must be kept at the beginning

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 2]
```diff
@@ -962,7 +962,9 @@ DBFAddNativeFieldType(DBFHandle psDBF, const char * pszFieldName,
 
         /* load record */
         psDBF->sHooks.FSeek( psDBF->fp, nRecordOffset, 0 );
-        psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp );
+        if (psDBF->sHooks.FRead(pszRecord, nOldRecordLength, 1, psDBF->fp) != 1) { 
+          return -1;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:966]: memory leak of pszRecord

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 3]
```diff
@@ -1899,7 +1901,9 @@ DBFDeleteField(DBFHandle psDBF, int iField)
 
         /* load record */
         psDBF->sHooks.FSeek( psDBF->fp, nRecordOffset, 0 );
-        psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp );
+        if (psDBF->sHooks.FRead(pszRecord, nOldRecordLength, 1, psDBF->fp) != 1) {
+          return FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:1905]: memleak of pszRecord

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 4]
```diff
@@ -2010,7 +2014,9 @@ DBFReorderFields( DBFHandle psDBF, int* panMap )
 
             /* load record */
             psDBF->sHooks.FSeek( psDBF->fp, nRecordOffset, 0 );
-            psDBF->sHooks.FRead( pszRecord, psDBF->nRecordLength, 1, psDBF->fp );
+            if (psDBF->sHooks.FRead(pszRecord, psDBF->nRecordLength, 1, psDBF->fp) != 1) {
+              return FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2018]: memleak of several things...

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 5]
```diff
@@ -2160,7 +2166,9 @@ DBFAlterFieldDefn( DBFHandle psDBF, int iField, const char * pszFieldName,
 
             /* load record */
             psDBF->sHooks.FSeek( psDBF->fp, nRecordOffset, 0 );
-            psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp );
+            if (psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp ) != 1) {
+              return FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2170]: memleak of several things

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 6]
```diff
@@ -2227,7 +2235,9 @@ DBFAlterFieldDefn( DBFHandle psDBF, int iField, const char * pszFieldName,
 
             /* load record */
             psDBF->sHooks.FSeek( psDBF->fp, nRecordOffset, 0 );
-            psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp );
+            if (psDBF->sHooks.FRead( pszRecord, nOldRecordLength, 1, psDBF->fp )!= 1) {
+              return FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2239]: memleak of several things

**Review by rouault — CHANGES_REQUESTED** [review]

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 1 above)
Comment (schwehr) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:314]: Really? :(. Is there any way to get it to C99?  20 years ago versus 30 years ago... I'll make an issue https://github.com/OSGeo/shapelib for discussion.

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 6 above)
Comment (schwehr) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2239]: Ah the joys of C.  I read right past all those issues.

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 1 above)
Comment (rouault) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:314]: I'm afraid I don't have any motivation/time/energy to discuss language updates for shapelib, for which I'm already a bad maintainer, especially for a one line change that can be easily avoided.

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 4 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2018]: Some fairly complex state being held here, I've reworked a bit to consolidate error handling, but I'm nervous that psDBF might be left in an invalid state after a partial loop through nRecords.  Thoughts on how to handle this correctly?

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 6 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2239]: Yeah, I've added some ad-hoc error handling, please let me know if there's a better approach.

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 1 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:314]: Moved declaration back up.

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 2 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:967]: Done

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 3 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:1907]: Done

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 7]
```diff
@@ -1995,6 +1997,8 @@ DBFReorderFields( DBFHandle psDBF, int* panMap )
     free(psDBF->pszHeader);
     psDBF->pszHeader = pszHeaderNew;
 
+    bool errorAbort = FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2000]: to move at beginning of function

Comment (rouault) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2000]: use int rather than bool for C89...

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c [hunk 8]
```diff
@@ -2150,6 +2166,7 @@ DBFAlterFieldDefn( DBFHandle psDBF, int iField, const char * pszFieldName,
     psDBF->bNoHeader = TRUE;
     DBFUpdateHeader( psDBF );
 
+    bool errorAbort = FALSE;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/shape/dbfopen.c:2169]: same remarks as above

**Review by rouault — CHANGES_REQUESTED** [review]

File: gdal/ogr/ogrsf_frmts/shape/dbfopen.c (hunk 8 above)
Comment (slacy) [reply gdal/ogr/ogrsf_frmts/shape/dbfopen.c]: done

**Review by rouault — APPROVED** [review]

**Pull request merged**